File: LeetCode/337. House Robber III/Solution.kt
package problem337

class Solution {
  fun rob(root: TreeNode?): Int {
    return buildRobTree(root)!!.`val`
  }

  /** return rob tree node */
  private fun buildRobTree(tree: TreeNode?): TreeNode? {
    if (tree == null) return null

    val rob = TreeNode(0)
    rob.left = buildRobTree(tree.left)
    rob.right = buildRobTree(tree.right)

    // 1. we don't take treeNode, take tree's child
    val count1 = sumChild(rob)
    // 2. we take treeNode, so we can't take tree's child, but take tree's grand child
    val count2 = tree.`val` + sumChild(rob.left) + sumChild(rob.right)

    rob.`val` = maxOf(count1, count2)

    return rob
  }

  private fun sumChild(node: TreeNode?): Int {
    return (node?.left?.`val` ?: 0) + (node?.right?.`val` ?: 0)
  }
}

class TreeNode(var `val`: Int) {
  var left: TreeNode? = null
  var right: TreeNode? = null
}

fun main() {

}

File: LeetCode/337. House Robber III/Solution.cc
#include <algorithm>  // for std::min
#include <vector>

using std::vector;
struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode() : val(0), left(nullptr), right(nullptr) {}
  TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
  TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
 public:
  int rob(TreeNode *root) { return robSub(root)->val; }

  TreeNode *robSub(TreeNode *tree) {
    if (tree == nullptr) return nullptr;

    auto rob = new TreeNode{};
    rob->left = robSub(tree->left);
    rob->right = robSub(tree->right);

    int unselect = sumChild(rob);
    int select = tree->val + sumChild(rob->left) + sumChild(rob->right);

    rob->val = std::max(unselect, select);

    return rob;
  }

  int sumChild(TreeNode *tree) {
    int sum = 0;
    if (tree != nullptr && tree->left != nullptr) {
      sum += tree->left->val;
    }
    if (tree != nullptr && tree->right != nullptr) {
      sum += tree->right->val;
    }
    return sum;
  }
};

// test

#include <iostream>

void test(const std::vector<int> &height, int expected) {
  int ret = Solution().trap(height);
  std::cout << ret << " == " << expected << std::endl;
}

int main() {
  test(std::vector<int>{4, 2, 0, 3, 2, 5}, 9);
  test(std::vector<int>{0, 1, 0, 2, 1, 0, 1, 3, 2, 1, 2, 1}, 6);
  test(std::vector<int>{0, 1, 0, 2, 1, 0, 1, 3, 2, 1, 2, 1}, 6);
}
